### PR TITLE
module: picard filtersamreads

### LINKED ISSUE #591: new module: PICARD/FILTERSAMREADS
I think it would be good to have a module for [PICARD/FILTERSAMREADS](https://bioconda.github.io/recipes/picard/README.html)

- [x] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [x] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [x] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [x] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### LINKED ISSUE #600: new module: PICARD/SORTSAM
I think it would be good to have a module for [PICARD/SORTSAM](https://bioconda.github.io/recipes/picard/README.html)

- [x] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [x] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [x] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [x] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by jfy133**
## PR checklist

Adds picard FilterSamReads

Closes #591 

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `<SOFTWARE>.version.txt` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

**Comment by jfy133** [comment]
TODO: replace integrated sortsam with #603 in testing instead

File: modules/picard/filtersamreads/main.nf [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    input:
+    tuple val(meta), path(bam)
+    val filter
+    path readlist
+
+    output:
+    tuple val(meta), path("*.filtered.bam"), emit: bam
+    path "*.version.txt"                   , emit: version
```
Comment (drpatelh) [inline modules/picard/filtersamreads/main.nf:28]:
```suggestion
    tuple val(meta), path("*.bam"), emit: bam
    path "*.version.txt"          , emit: version
```

File: modules/picard/filtersamreads/main.nf [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    picard \\
+        FilterSamReads \\
+        -Xmx${avail_mem}g \\
+        --INPUT $bam \\
+        --OUTPUT ${prefix}.filtered.bam \\
+        --FILTER $filter \\
+        $options.args
+
+    echo \$(picard FilterSamReads --version 2>&1) | grep -o 'Version:.*' | cut -f2- -d: > ${software}.version.txt
+    """
```
Comment (drpatelh) [inline modules/picard/filtersamreads/main.nf:50]:
```suggestion
    if ( filter == 'includeAligned' || filter == 'excludeAligned' ) {
        """
        picard \\
            FilterSamReads \\
            -Xmx${avail_mem}g \\
            --INPUT $bam \\
            --OUTPUT ${prefix}.filtered.bam \\
            --FILTER $filter \\
            $options.args

        echo \$(picard FilterSamReads --version 2>&1) | grep -o 'Version:.*' | cut -f2- -d: > ${software}.version.txt
        """
```

File: modules/picard/filtersamreads/main.nf [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        FilterSamReads \\
+        -Xmx${avail_mem}g \\
+        --INPUT $bam \\
+        --OUTPUT ${prefix}.filtered.bam \\
+        --FILTER $filter \\
+        --READ_LIST_FILE $readlist \\
+        $options.args
+
+    echo \$(picard FilterSamReads --version 2>&1) | grep -o 'Version:.*' | cut -f2- -d: > ${software}.version.txt
+    """
```
Comment (drpatelh) [inline modules/picard/filtersamreads/main.nf:63]:
```suggestion
    } else if ( filter == 'includeReadList' || filter == 'excludeReadList' ) {
		"""
		picard \\
		    FilterSamReads \\
		    -Xmx${avail_mem}g \\
		    --INPUT $bam \\
		    --OUTPUT ${prefix}.bam \\
		    --FILTER $filter \\
		    --READ_LIST_FILE $readlist \\
		    $options.args

		echo \$(picard FilterSamReads --version 2>&1) | grep -o 'Version:.*' | cut -f2- -d: > ${software}.version.txt
		"""
```

Comment (drpatelh) [inline modules/picard/filtersamreads/main.nf:63]:
```suggestion
"""
		picard \\
		    FilterSamReads \\
		    -Xmx${avail_mem}g \\
		    --INPUT $bam \\
		    --OUTPUT ${prefix}.filtered.bam \\
		    --FILTER $filter \\
		    --READ_LIST_FILE $readlist \\
		    $options.args

		echo \$(picard FilterSamReads --version 2>&1) | grep -o 'Version:.*' | cut -f2- -d: > ${software}.version.txt
		"""
```

File: tests/config/test_data.config [hunk 4]
```diff
@@ -74,12 +74,14 @@ params {
                 scaffolds_fasta                                = "${test_data_dir}/genomics/sarscov2/illumina/fasta/scaffolds.fasta"
 
                 assembly_gfa                                   = "${test_data_dir}/genomics/sarscov2/illumina/gfa/assembly.gfa"
+
+                picard_readlist_test_single_end_bam            = "${test_data_dir}/genomics/sarscov2/illumina/picard/test.single_end.bam.readlist.txt"
```
Comment (drpatelh) [inline tests/config/test_data.config:78]:
```suggestion
                assembly_gfa                                   = "${test_data_dir}/genomics/sarscov2/illumina/gfa/assembly.gfa"

                test_single_end_bam_readlist_txt               = "${test_data_dir}/genomics/sarscov2/illumina/picard/test.single_end.bam.readlist.txt"
```

Comment (drpatelh) [reply tests/config/test_data.config:78]: The name is always the same as the final file but dots replaced with underscores

File: tests/modules/picard/filtersamreads/main.nf [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+include { PICARD_FILTERSAMREADS } from '../../../../modules/picard/filtersamreads/main.nf' addParams( options: [:] )
+
+workflow test_picard_filtersamreads {
+
+    input = [ [ id:'test', single_end:false ], // meta map
+              file(params.test_data['sarscov2']['illumina']['test_single_end_bam'], checkIfExists: true) ]
+    sort_order = 'queryname'
+    filter = 'includeAligned'
+
+    PICARD_SORTSAM ( input, sort_order  )
```
Comment (drpatelh) [inline tests/modules/picard/filtersamreads/main.nf:15]:
```suggestion
    PICARD_SORTSAM ( input, sort_order )
```

File: tests/modules/picard/filtersamreads/test.yml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+- name: picard filtersamreads
+  command: nextflow run ./tests/modules/picard/filtersamreads -entry test_picard_filtersamreads -c tests/config/nextflow.config
+  tags:
+    - picard
+    - picard/filtersamreads
+  files:
+    - path: output/picard/test.filtered.bam
+      md5sum: b44a6ca04811a9470c7813c3c9465fd5
+
```
Comment (drpatelh) [inline tests/modules/picard/filtersamreads/test.yml:9]:
```suggestion
```

File: modules/picard/filtersamreads/main.nf (hunk 1 above)
Comment (jfy133) [reply modules/picard/filtersamreads/main.nf:28]: Wouldn't this also pick up the input `bam`?

File: tests/modules/picard/filtersamreads/main.nf [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    PICARD_FILTERSAMREADS ( PICARD_SORTSAM.out.bam, filter, [] )
+}
+
+workflow test_picard_filtersamreads_readlist {
+
+    input = [ [ id:'test', single_end:false ], // meta map
+              file(params.test_data['sarscov2']['illumina']['test_single_end_bam'], checkIfExists: true) ]
+    filter = 'includeReadList'
+    readlist = file(params.test_data['sarscov2']['illumina']['picard_readlist_test_single_end_bam'], checkIfExists: true)
```
Comment (jfy133) [inline tests/modules/picard/filtersamreads/main.nf:24]:
```suggestion
    readlist = file(params.test_data['sarscov2']['illumina']['test_single_end_bam_readlist_txt'], checkIfExists: true)
```

File: modules/picard/filtersamreads/main.nf (hunk 1 above)
Comment (drpatelh) [reply modules/picard/filtersamreads/main.nf:28]: Yup. But you need to change the filename via `suffix`. We need to change this too https://github.com/nf-core/modules/blob/f43778b0e641a9643ae73fe1f2fe2d5bb81edffc/modules/picard/sortsam/main.nf#L44. Missed it.

File: tests/modules/picard/filtersamreads/main.nf [hunk 8]
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { PICARD_SORTSAM        } from '../../../../modules/picard/sortsam/main.nf'        addParams( options: [:] )
+include { PICARD_FILTERSAMREADS } from '../../../../modules/picard/filtersamreads/main.nf' addParams( options: [:] )
```
Comment (drpatelh) [inline tests/modules/picard/filtersamreads/main.nf:6]:
```suggestion
include { PICARD_FILTERSAMREADS } from '../../../../modules/picard/filtersamreads/main.nf' addParams( options: [suffix:'.filtered'] )
```

File: modules/picard/filtersamreads/main.nf (hunk 1 above)
Comment (drpatelh) [reply modules/picard/filtersamreads/main.nf:28]: Like https://github.com/nf-core/modules/pull/602/files#r673951434

Comment (drpatelh) [reply modules/picard/filtersamreads/main.nf:28]: This allows developers to call their output files whatever they want.

Comment (drpatelh) [reply modules/picard/filtersamreads/main.nf:28]: May be worth fixing the `picard/sortsam` module first :)

File: modules/picard/filtersamreads/meta.yml [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+output:
+    - meta:
+        type: map
+        description: |
+            Groovy Map containing sample information
+            e.g. [ id:'test', single_end:false ]
+    - bam:
+        type: file
+        description: Filtered BAM file
+        pattern: "*.filtered.{bam}"
```
Comment (jfy133) [inline modules/picard/filtersamreads/meta.yml:44]:
```suggestion
        pattern: "*.{bam}"
```

File: modules/picard/filtersamreads/main.nf [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+            FilterSamReads \\
+            -Xmx${avail_mem}g \\
+            --INPUT $bam \\
+            --OUTPUT ${prefix}.bam \\
+            --FILTER $filter \\
+            --READ_LIST_FILE $readlist \\
+            $options.args
+
+            echo \$(picard FilterSamReads --version 2>&1) | grep -o 'Version:.*' | cut -f2- -d: > ${software}.version.txt
+        """
```
Comment (drpatelh) [inline modules/picard/filtersamreads/main.nf:63]:
```suggestion
        echo \$(picard FilterSamReads --version 2>&1) | grep -o 'Version:.*' | cut -f2- -d: > ${software}.version.txt
        """
```

**Review by drpatelh — APPROVED** [review]

**Pull request merged**